BUG: Fix segfault in np.random.shuffle for arrays of different length strings

simongibbons opened the pull request:
np.random.shuffle will allocate a buffer based on the size of the first
element of an array of strings. If the first element is smaller than
another in the array this buffer will overflow, causing a segfault
when garbage is collected.

This fixes the issue by ensuring the buffer is allocated based upon
the dtype of the array rather than just that of the first element.

Fixes #7710

simongibbons committed:
BUG: Fix segfaults in np.random.shuffle

np.random.shuffle will allocate a buffer based on the size of the first
element of an array of strings. If the first element is smaller than
another in the array this buffer will overflow, causing a segfault
when garbage is collected.

Additionally if the array contains objects then one would be left
in the buffer and have it's refcount erroniously decrimented on
function exit, causing that object to be deallocated too early.

To fix this we change the buffer to be an array of int8 of the
the size of the array's dtype, which sidesteps both issues.

Fixes #7710

rkern commented:
Good catch. LGTM.

seberg commented:
Maybe it would be better to use arr[0, ...] instead? Avoid the roundtrip to potential scalars?

simongibbons commented:
@seberg you're not wrong, a quick benchmark makes it look like that is ~6x faster!

mhvk commented:
I'd use `arr[:1]` instead, which I think is a bit clearer (the `...` suggest trailing dimensions that may not be there, so to me it needs a bit more braintwisting to see why this is useful). But really this is nitpicking!

jaimefrio commented:
What we should have for sure is a comment in the code explaining why this is needed.

simongibbons commented:
@jaimefrio Yes I agree, whilst the code was self explanatory when using the `dtype` argument it now certainly needs a comment.

@mhvk I think that version is easier on the old grey matter as well!

seberg commented on numpy/random/tests/test_regression.py:
```diff
@@ -113,5 +113,20 @@ def test_choice_sum_of_probs_tolerance(self):
+        for _ in range(100):
+            z = np.random.permutation(a)
+
+        # Force Garbage Collection - should not segfault.
```
Is explicit garbage collection really necessary? Everything seems like it should be cleaned up on function exit anyway (no obvious reference loops or so?).

simongibbons replied:
The explicit garbage collection is to trigger the segfault within the test rather than at some arbitrary time in the future (it could even be when the interpreter exits). Thus if this test fails sometime in the future we will know what the failure is!

charris replied:
@seberg This is more deterministic.I think function exit just decrefs the objects and the gc will run whenever it feels like it, which is why the original function call tended to segfault on Python exit.

seberg replied:
Hmm, ok.

charris replied:
Small array data, IIRC, comes off a cache, so that may have something to do with it.

seberg commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -5064,7 +5064,8 @@ cdef class RandomState:
             itemsize = x.dtype.itemsize
-            buf = np.empty_like(x[0])  # GC'd at function exit
+            # Ensure that buf has the same dtype as x
+            buf = np.empty_like(x[:1])  # GC'd at function exit
```
OK, I think the object cases were probably already all fine (though I am not 100% sure, but oh wlel...)

charris replied:
Indexing speaks more loudly to you than to the rest of us ;) Agree about the comment, more explanation is needed.

charris replied:
Something like
```
# The scalar x[0] has a dtype just big enough to hold that string, so use x[:1] to get a # dtype big enough to hold any string in x.

mhvk replied:
Maybe add at start: "If x is one-dimensional, x[0] would be a scalar with a dtype just ..."

charris replied:
Yeah `a[0, ...]` is a bit cleaner as it doesn't have an extra first dimension.

simongibbons replied:
I agree with both of you with the comment issue. I'll update it.

However @seberg you are right to be worried about the object case, looks like it was working before (in 1.11) but this fix breaks it for a reason that is not entirely obvious to me just yet.

seberg replied:
@simongibbons is there a bug, can't figure it out right now? For the multi-dimensional case, nothing should change as far as I can see. For the 1D case, it seems to me things may be broken if the dtype of the result of `arr[0]` (the scalar) is smaller then an object pointer, the buffer is too small, but this change should fix that as well.

simongibbons replied:
Yes there is a bug for the object case

    a = np.array([[np.arange(1), np.arange(53)], np.arange(241)])

    for _ in range(100):
        z = np.random.permutation(a)

Will segfault in both the old and new versions.

You're right in saying that the buffer is the right size for an object pointer in this case, so that can't be the cause.

simongibbons replied:
@seberg  Yes there is a bug for the object case

    a = np.array([[np.arange(1), np.arange(53)], np.arange(241)])

    for _ in range(100):
        z = np.random.permutation(a)

Will segfault in both the old and new versions.

You're right in saying that the buffer is the right size for an object pointer in this case, so that can't be the cause.

charris replied:
Probably a reference counting bug.

seberg replied:
Ah, of course, the view will incref its elements, then they get shuffled under its feet and the decref goes to the wrong one. Easiest solution is likely to just avoid that 1D optimization for object arrays by checking `arr.dtype.hasobject` first.

charris replied:
Whoa, well sighted, definitely needs a comment as well.

charris replied:
Bit subtle, but `buf` is left containing a reference to some object in `x`, which is decrefed when the buffer is destroyed. Furthermore, `buf` begins containing a reference to `None`, so there is also a memory leak. The only other fix I can see to define the buffer as `int8` with length `itemsize`.

seberg replied:
Ah, that is maybe more elegant. Should probably make sure the raw
shuffle does not release the GIL then (no idea if it does).

simongibbons commented:
@charris Yes you're right that was a subtle one to track down. Very nice spot. I also like your solution, it's very elegant.

I've implemented that now and added a test for the object case.

Given that all of the swaps in the multidimensional case are done in python rather than doing memcpys so the object left in the buffer there will be correctly refcounted and the buffer was picking up the correct dtype anyway. I've left that part of the code unchanged.

charris commented:
Thanks @simongibbons .

seberg commented:
Thanks!
